### DISCUSSION TITLE
Sqlalchemy Core create table with UniqueConstraint and condition

### DISCUSSION THREAD

**Discussion by Lepiloff** [sqlalchemy/sqlalchemy#8679] [Usage Questions] [answered]
**DB=postgresql**

there is a table created using this syntax

    event_invites = Table(
        "event_invites",
        metadata,
        Column("id", Integer(), primary_key=True),
        Column("from_user", ForeignKey(users.c.id, ondelete="SET NULL"), nullable=False),
        Column("to_user", ForeignKey(users.c.id, ondelete="SET NULL"), nullable=False),
        Column("to_event", ForeignKey(events.c.id, ondelete="CASCADE"), nullable=False),
        Column(
            "is_active",
            Boolean(),
            server_default=sql.expression.true(),
            nullable=True,
        ),
        ....
        ....
        UniqueConstraint('from_user', 'to_user', 'to_event', 'is_active', name='unique_event_invites')
    )

I need to add it so that the uniqueness of the fields is only triggered if the `is_active` field is `True`.

I have found a way to add this condition when creating tables with syntax for Sqlalchemy ORM

example

    class ExampleTable(Base):
    
        __tablename__ = 'example_table'
        __table_args__ = (
            Index(
                'ix_unique_primary_content',  # Index name
                'object_type', 'object_id',  # Columns which are part of the index
            unique=True,
            postgresql_where=Column('is_primary')),  # The condition
        )
    
        id = Column(Integer, primary_key=True)
        object_type = Column(Unicode(50))
        object_id = Column(Integer)
        is_primary = Column(Boolean)

But I can't figure out how to adapt it for my case

Something like this

    UniqueConstraint('from_user', 'to_user', 'to_event', 'is_active', name='unique_event_invites',
                     postgresql_where=Column("is_active=True")
                     )

Return erorr

    sqlalchemy.exc.ArgumentError: Argument 'postgresql_where' is not accepted by dialect 'postgresql' on behalf of <class 'sqlalchemy.sql.schema.UniqueConstraint'>

**Comment by zzzeek** [comment, marked as answer]
I dont know that PostgreSQL supports WHERE for UNIQUE constraints:

https://www.postgresql.org/docs/current/sql-altertable.html

```
and table_constraint is:

[ CONSTRAINT constraint_name ]
{ CHECK ( expression ) [ NO INHERIT ] |
  UNIQUE [ NULLS [ NOT ] DISTINCT ] ( column_name [, ... ] ) index_parameters |
  PRIMARY KEY ( column_name [, ... ] ) index_parameters |
  EXCLUDE [ USING index_method ] ( exclude_element WITH operator [, ... ] ) index_parameters [ WHERE ( predicate ) ] |
  FOREIGN KEY ( column_name [, ... ] ) REFERENCES reftable [ ( refcolumn [, ... ] ) ]
    [ MATCH FULL | MATCH PARTIAL | MATCH SIMPLE ] [ ON DELETE referential_action ] [ ON UPDATE referential_action ] }
[ DEFERRABLE | NOT DEFERRABLE ] [ INITIALLY DEFERRED | INITIALLY IMMEDIATE ]
```

so...why not use a unique index like you have ?

**Reply by CaselIT** [reply]
I confirm that you have to use an unique index, not an unique constraint for this.

Also you probably want the lowercase `column`, not `Column` in your example

**Comment by Lepiloff** [comment]
Thanks for reply
I can't find an example of how to add an Index to a table using SqlAlchemy Core. 
Only UniqueConstraint
https://docs.sqlalchemy.org/en/14/core/constraints.html#unique-constraint
Could you give me an example?

@CaselIT I'm using `Column`  as described in the documentation https://docs.sqlalchemy.org/en/14/core/constraints.html#defining-foreign-keys

**Reply by CaselIT** [reply]
>  I can't find an example of how to add an Index to a table using SqlAlchemy Core.

The documentation of Index has numerous examples https://docs.sqlalchemy.org/en/14/core/constraints.html?highlight=index#sqlalchemy.schema.Index

Regarding the the upper case `Column` is in general used only when passed to a `Table`.

For one off columns and tables to use in sql construct you can use the lowercase versions [column](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=column#sqlalchemy.sql.expression.column) [table](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.table)

but if it's working it should be fine in this case
